src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java

```diff
@@ -0,0 +1,176 @@
+import com.ctre.phoenix.motorcontrol.can.TalonFX;
+import com.ctre.phoenix.motorcontrol.can.TalonSRX;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+import frc.robot.Ports;
+import frc.robot.UnitModel;
+import frc.robot.Utils;
+
+public class SwerveModule extends SubsystemBase {
+    private final TalonFX driveMotor;
+    private final TalonSRX angleMotor;
+    private final int wheel;
+
+    private final UnitModel unitDrive = new UnitModel(Constants.SwerveDrive.TICKS_PER_METER);
+    private final UnitModel unitAngle = new UnitModel(Constants.SwerveDrive.TICKS_PER_RAD);
+
+    public SwerveModule(int wheel, int driveMotorPort, int angleMotorPort, boolean[] inverted) {
+
+        driveMotor = new TalonFX(driveMotorPort);
+        angleMotor = new TalonSRX(angleMotorPort);
+
+        // configure feedback sensors
+        angleMotor.configSelectedFeedbackSensor(FeedbackDevice.Analog, 0, Constants.TALON_TIMEOUT);
+        angleMotor.configFeedbackNotContinuous(Ports.SwerveDrive.IS_NOT_CONTINUOUS_FEEDBACK, Constants.TALON_TIMEOUT);
+
+        angleMotor.setNeutralMode(NeutralMode.Brake);
+
+        // set inversions
+        angleMotor.setInverted(inverted[0]);
+        driveMotor.setInverted(inverted[1]);
+
+        angleMotor.setSensorPhase(inverted[2]);
+        driveMotor.setSensorPhase(inverted[3]);
+
+        // Set amperage limits
+        SupplyCurrentLimitConfiguration currLimitConfig = new SupplyCurrentLimitConfiguration(
+                Constants.ENABLE_CURRENT_LIMIT,
+                Constants.SwerveDrive.MAX_CURRENT,
+                Constants.SwerveModule.TRIGGER_THRESHOLD_CURRENT,
+                Constants.SwerveModule.TRIGGER_THRESHOLD_TIME
+        );
+
+        driveMotor.configSupplyCurrentLimit(currLimitConfig);
+
+        angleMotor.configContinuousCurrentLimit(Constants.SwerveDrive.MAX_CURRENT);
+        angleMotor.enableCurrentLimit(Constants.ENABLE_CURRENT_LIMIT);
+
+        configPIDF();
+
+        // set voltage compensation and saturation
+        driveMotor.enableVoltageCompensation(Constants.SwerveModule.ENABLE_VOLTAGE_COMPENSATION);
+        driveMotor.configVoltageCompSaturation(Constants.SwerveModule.VOLTAGE_SATURATION);
+
+        angleMotor.enableVoltageCompensation(Constants.SwerveModule.ENABLE_VOLTAGE_COMPENSATION);
+        angleMotor.configVoltageCompSaturation(Constants.SwerveModule.VOLTAGE_SATURATION);
+
+        angleMotor.selectProfileSlot(0, 0);
+        driveMotor.selectProfileSlot(1, 0);
+
+        this.wheel = wheel;
+    }
+
+
+    /**
+     * @return the speed of the wheel in [m/s]
+     */
+    public double getSpeed() {
+        return unitDrive.toVelocity(driveMotor.getSelectedSensorVelocity(1));
+    }
+
+    /**
+     * @return the angle of the wheel in radians
+     */
+    public double getAngle() {
+        return unitAngle.toUnits(angleMotor.getSelectedSensorPosition() - Constants.SwerveModule.ZERO_POSITION[wheel]);
```

Comment:
> `- Constants.SwerveModule.ZERO_POSITION[wheel]`

If I understand the intention correctly, you want to compensate for the encoder offset. Won't it be simpler to call `angleMotor.setSelectedSensorPosition()` with the "zero position" in this class' constructor?

Reply:
Yes it would be